&yarnabrina [ENH] `skforecast` ForecasterAutoreg adapter

Abhay-Lejith opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->
Continuing with work done in #5447 .

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
An adapter for the skforecast `ForecasterAutoreg` .
#### Does your contribution introduce a new dependency? If yes, which one?
skforecast
<!--
Only relevant if you changed pyproject.toml.
We try to minimize dependencies in the core dependency set. There
are also further specific instructions to follow for soft dependencies.
See here for handling dependencies in sktime: https://www.sktime.net/en/latest/developer_guide/dependencies.html
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?
Yes.
<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

<!--
Thanks for contributing!
-->

Abhay-Lejith commented:
@fkiraly @yarnabrina regarding the index error that was occurring here. skforecast expects only `pd.DatetimeIndex` or `pd.RangeIndex`. I saw in the discussion in #5447 that setting the `enforce_index_type` tag is not related to any logic which enforces it. 
So now, we can either check index type in `fit` and return an error saying that the particular index type is not supported or we can try to make the conversion ourselves.

yarnabrina commented:
Have you seen #5997? May be something like this will help.

https://github.com/sktime/sktime/blob/fd2ebc1ab5fd4b543a9be41e36349d8a96f2ea6c/sktime/forecasting/compose/_darts_reduce.py#L309-L337

Abhay-Lejith commented:
> Have you seen #5997? May be something like this will help.

I see, so you suggest making a function to convert index types.
Since this is something which could be used in other places as well, I suppose it would be better to write a function separately in like utils maybe and import and call it wherever required? 
What are your thoughts @yarnabrina @fkiraly ?

yarnabrina commented:
> Since this is something which could be used in other places as well, I suppose it would be better to write a function separately in like utils maybe and import and call it wherever required?

I think this is a good idea in principal. However, few of the packages tend to have specific requirements and we have to change slightly to suit the purpose. I think similar functions exist for existing (merged) adapters already, and you can compare to notice slight differences.

If we can come up with a generalised conversion function (I am thinking private - will it be useful as public function 🤔 ), that'd be great IMO 👍🏽

Abhay-Lejith commented:
https://github.com/sktime/sktime/blob/b429fd939539b39286c38e410ef2feab374f55c6/sktime/utils/_testing/forecasting.py#L131-L135

This function in tests is giving me a lot of problems. The `expected` index type keeps changing in each test case (sometimes its `pd.DatetimeIndex` , sometimes its `pd.PeriodIndex` etc) , I am not able to figure out on what basis it gets set.
I'm getting errors because the index types don't match( the values themselves are correct).
@fkiraly @yarnabrina , some help here would be great.

* If I can know what the index type of `expected` will be, I can change the index type of `y_pred_index` accordingly. 
* Also, shouldn't the `expected` index type simply be the same as that of whatever input was passed into fit/predict ?

fkiraly commented:
> Since this is something which could be used in other places as well, I suppose it would be better to write a function separately in like utils maybe and import and call it wherever required?
What are your thoughts @yarnabrina @fkiraly ?

I completely agree! This is something we have been considering for a while, given that the `ForecastingHorizon` needs a rework. See for instance https://github.com/sktime/sktime/issues/6546

There are two ways one could approach this: bottom-up, having it first in a few concrete estimators an then refactor it to a joint utils; or top-down, starting with a design document and then implementing and integrating. Bottom-up is usually easier if you start getting into a code base. I agree with @yarnabrina that having sth like this in a generalizable way would be very useful.

> If I can know what the index type of `expected` will be

I would say, the trick is to use `fh` directly, namely `ForecastingHorizon.get_expected_pred_idx` (or `to_absolute_index`). There is no requirement that you reimplement the index generation logic in `_predict`.

> Also, shouldn't the expected index type simply be the same as that of whatever input was passed into fit/predict?

Yes, it should be (is it not?), but that's just the type. The test also checks for values.

If you think the logic of the test is broken, kindly flag that.

yarnabrina commented on pyproject.toml:
```diff
@@ -101,6 +101,7 @@ all_extras = [
   "scikit_posthocs>=0.6.5",
   "seaborn>=0.11",
   "seasonal",
+  'skforecast<=0.12.1,>=0.11; python_version < "3.12"',
```
This specification is not recommended. It's format in general should be greater than or equal to (inclusive) minimum (officially) supported stable (usually minor) version, and less than (exclusive) next upcoming stable version (again, usually minor).

Since we are adding it now, we should use latest version as minimum.

```suggestion
  'skforecast<0.13,>=0.12.1; python_version < "3.12"',
```

yarnabrina replied:
@fkiraly / @benHeid please share your comments here too.

Abhay-Lejith replied:
Understood, I've made this change in my latest commit.

yarnabrina commented on pyproject.toml:
```diff
@@ -138,6 +139,7 @@ all_extras_pandas2 = [
   "scikit_posthocs>=0.6.5",
   "seaborn>=0.11",
   "seasonal",
+  'skforecast<=0.12.1,>=0.11; python_version < "3.12"',
```
```suggestion
  'skforecast<0.13,>=0.12.1; python_version < "3.12"',
```

yarnabrina commented on pyproject.toml:
```diff
@@ -183,6 +185,7 @@ forecasting = [
   "arch>=5.6,<7.1",
   'pmdarima!=1.8.1,<2.1,>=1.8; python_version < "3.12"',
   "prophet<1.2,>=1.1",
+  'skforecast<=0.12.1,>=0.11; python_version < "3.12"',
```
```suggestion
  'skforecast<0.13,>=0.12.1; python_version < "3.12"',
```

yarnabrina commented on sktime/forecasting/compose/__init__.py:
```diff
@@ -29,7 +29,9 @@
     "YfromX",
+    "SkforecastAutoreg",
     "IgnoreX",
+    "SkforecastAutoreg",
```
Why is this repeated? 😕

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,380 @@
+        weight_func: typing.Optional[typing.Callable] = None,
+        differentiation: typing.Optional[int] = None,
+        fit_kwargs: typing.Optional[dict] = None,
+    ) -> None:
```
As per the [docs](https://skforecast.org/0.12.1/api/forecasterautoreg), they support `binner_kwargs` too.

Abhay-Lejith replied:
Will check it out.

Abhay-Lejith replied:
I've added all missing args now.

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,380 @@
+        "capability:pred_int": True,
+        "capability:pred_int:insample": False,
+        "python_version": ">=3.8,<3.13",
+        "python_dependencies": ["skforecast>=0.11"],
```
```suggestion
        "python_dependencies": ["skforecast>=0.12.1"],
```

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,380 @@
+        )
+
+    @staticmethod
+    def _coerce_column_names(X: pandas.DataFrame):
```
This type hint implies `X` can only be a `pandas.DataFrame`. If you want to allow `None`, please use `typing.Optional`.

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,380 @@
+        from sklearn.linear_model import LinearRegression
+
+        params = [
+            {"regressor": LinearRegression(), "lags": 2},
+            {"regressor": RandomForestRegressor(), "lags": [1, 3]},
```
We should probably add coverage for more arguments.

Abhay-Lejith replied:
Will do once the existing errors are resolved.

fkiraly replied:
well, they now are :-)

yarnabrina commented:
> The expected index type keeps changing in each test case (sometimes its pd.DatetimeIndex , sometimes its pd.PeriodIndex etc)

> shouldn't the expected index type simply be the same as that of whatever input was passed into fit/predict ?

You answered yourself 😄 I have to see the errors you are getting, but most likely what is happening is that this test gets run multiple times with different values of the fixture, and based on the index type the expectation keeps changing.

Abhay-Lejith commented:
So I've fixed one of the index issues. I've implemented two index conversion functions, one to convert `integer` index to `RangeIndex` and another to convert `PeriodIndex` to `DateTimeIndex`. 
I've spotted the reason for one of the other index issues as well. In cases where vectorized fit is called, only X index is getting converted and y index isn't which causes it to fail the X,y checks because indexes dont match anymore. Ideally once it passes through the private fit method, both should get converted. But in the vectorized version, there is some inconsistency. Is the y modified  in private fit not used by the vectorized fit calls? I'm having trouble tracing it through all the vectorized and parallel func calls.

Abhay-Lejith commented:
Looks like all issues but one have been solved. The remaining issue is with `predict_quantiles` and more specifically, ONLY in the case where `y` is `univariate` and we have a list of alpha. No idea how this is happening, I compared the output with ARIMA to see if there was any difference but there was none. 

This is the failure point:
https://github.com/sktime/sktime/blob/1b45876602f8116906edcf98d3942280791a3e68/sktime/datatypes/_proba/_check.py#L100-L101

This is the error:
```E           ValueError: could not broadcast input array from shape (2,2) into shape (1,2)```
This is happening in the obj.isna().values call.


I ran this code to see what is the format of obj when using another estimator(ARIMA) for which the test does not fail.
```python
from sktime.utils.estimator_checks import check_estimator
from sktime.forecasting.compose import SkforecastAutoreg
from sktime.forecasting.arima import ARIMA
check_estimator(ARIMA,fixtures_to_run='test_predict_quantiles[ARIMA-0-y:1cols-fh=1-alpha=[0.25, 0.75]]', raise_exceptions=True)
check_estimator(SkforecastAutoreg,fixtures_to_run='test_predict_quantiles[SkforecastAutoreg-0-y:1cols-fh=1-alpha=[0.25, 0.75]]', raise_exceptions=True)

```
This was the output ( look at the printed `obj` and details , the first one is from `ARIMA` call and the next one is from `SkforecastAutoreg` call ):
```
obj:
                    0          
                0.25      0.75
2000-02-20  2.343876  3.482099
obj index: DatetimeIndex(['2000-02-20'], dtype='datetime64[ns]', freq=None)
columns: MultiIndex([(0, 0.25),
            (0, 0.75)],
           )
All tests PASSED!

```
`ARIMA` test for the same fixture passed as seen above.
`SkforecastAutoReg` test for same fixture failed as seen below. ( Notice that `obj` is exactly the same in both cases? yet it fails only here?)

```
obj:
                    0          
                0.25      0.75
2000-02-20  2.167503  2.167503
obj index: DatetimeIndex(['2000-02-20'], dtype='datetime64[ns]', freq=None)
columns: MultiIndex([(0, 0.25),
            (0, 0.75)],
           )
Traceback (most recent call last):
  File "/home/abhay/trial.py", line 5, in <module>
    check_estimator(SkforecastAutoreg,fixtures_to_run='test_predict_quantiles[SkforecastAutoreg-0-y:1cols-fh=1-alpha=[0.25, 0.75]]', raise_exceptions=True)
  File "/home/abhay/sktime/sktime/utils/estimator_checks.py", line 116, in check_estimator
    test_cls_results = test_cls().run_tests(
  File "/home/abhay/sktime/sktime/tests/test_all_estimators.py", line 608, in run_tests
    test_fun(**deepcopy(args))
  File "/home/abhay/sktime/sktime/forecasting/tests/test_all_forecasters.py", line 507, in test_predict_quantiles
    self._check_predict_quantiles(
  File "/home/abhay/sktime/sktime/forecasting/tests/test_all_forecasters.py", line 433, in _check_predict_quantiles
    valid, msg, _ = check_is_mtype(
  File "/home/abhay/sktime/sktime/datatypes/_check.py", line 189, in check_is_mtype
    res = check_dict[key](obj, return_metadata=return_metadata, var_name=var_name)
  File "/home/abhay/sktime/sktime/datatypes/_proba/_check.py", line 104, in check_pred_quantiles_proba
    metadata["has_nans"] = obj.isna().values.any()
  File "/home/abhay/anaconda3/envs/sktime-dev/lib/python3.10/site-packages/pandas/core/frame.py", line 12664, in values
    return self._mgr.as_array()
  File "/home/abhay/anaconda3/envs/sktime-dev/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 1694, in as_array
    arr = self._interleave(dtype=dtype, na_value=na_value)
  File "/home/abhay/anaconda3/envs/sktime-dev/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 1753, in _interleave
    result[rl.indexer] = arr
ValueError: could not broadcast input array from shape (2,1) into shape (1,1)
```
It is clear that obj in both cases is the exact same, but in the case of skforecast it fails.

yarnabrina commented:
Assuming @Abhay-Lejith commented after his last commit is pushed and CI has progressed somewhat, it was 12 hours ago from now. @fkiraly any idea why 49 jobs are still queued? What's running elsewhere?

---

I don't have an answer to the question, but I'd like to see this `metadata` object and whether it varies between ARIMA and SkforecastAutoReg.

Another question of interest will be what regressor is skforecast adapter using here? Does that choice have any effect? Probably not seeing your `obj` values, but want to know.

Another comment is I note `0.25` and `0.75` quantiles seem same. That's definitely possible, but just pointing it out in case something else is wrong without causing failures.

Abhay-Lejith commented:
> but I'd like to see this metadata object and whether it varies between ARIMA and SkforecastAutoReg.

@yarnabrina the issue is when the value of the metadata is being calculated. i.e when `obj.isna().values.any()` is called. More specifically, it breaks when it reaches the `values` call. Looking at how obj is the same in both examples as I showed above, I do not know what could be the issue and am stuck.

yarnabrina commented:
Sorry if it sounds obvious, but my probably stupid question is if you just run right hand side of that line, i.e. only calculation of count of missing values, does it fail? Or does it fail only when you try to assign that to metadata object?

fkiraly commented:
I have not run the test, but my inspection of the traceback did not highlight anything obvious:

* `obj` seems to be a `pd.DataFrame` with columns as expected
* in particular, no index duplications
* it looks like the RHS alreay causes the error, i.e., `isna.any`

I have no idea why it should.

Probably there is something with the specific subtype of `DataFrame` or `index`, so I would take `ARIMA` and this forecaster, and manually inspect every aspect of `obj` when it reaches that line: entries, indices, index types, and `dtype`-s.

Abhay-Lejith commented:
> Probably there is something with the specific subtype of DataFrame or index, so I would take ARIMA and this forecaster, and manually inspect every aspect of obj when it reaches that line: entries, indices, index types, and dtype-s.

I already did this , it is printed below obj in this comment https://github.com/sktime/sktime/pull/6531#issuecomment-2159910778
All other details like index,columns,dtype is matching.

yarnabrina commented:
> Only case thats failing is univariate y with list of alpha

@Abhay-Lejith can you please tell how are you inferring this? I'm confused seeing lines 8287, 8284, 8283, etc. in https://github.com/sktime/sktime/actions/runs/9460646640/job/26064980911?pr=6531.

Abhay-Lejith commented:
> Only case thats failing is univariate y with list of alpha

This was with respect to the test `test_predict_quantiles`.
The Skforecast adapter only has a predict quantile method. So I assumed that the other failures like those happening in `test_predict_interval` (these are the ones you mentioned) will be resolved if we resolve predict_quantile. (since predict interval uses the predict quantile method).

fkiraly reviewed (changes requested):
@Abhay-Lejith, just to make it clear what imo the next step is, it is debugging the failure in `_check_predict_intervals`. What is the last objects seen in `sktime` before the failure, and how does it differ from the same test in `ARIMA`.

I recommend to use step-through or breakpoint debugging for this, as you may need to inspect the objects manually, e.g., for `dtype` or size.

Abhay-Lejith commented:
> > Probably there is something with the specific subtype of DataFrame or index, so I would take ARIMA and this forecaster, and manually inspect every aspect of obj when it reaches that line: entries, indices, index types, and dtype-s.
> 
> I already did this , it is printed below obj in this comment [#6531 (comment)](https://github.com/sktime/sktime/pull/6531#issuecomment-2159910778) All other details like index,columns,dtype is matching.

@fkiraly I already checked with breakpoint debugging and have posted the results above. Everything is matching.

fkiraly commented:
Hm, curious. Have you done what I suggested twice?

> All other details like index,columns,dtype is matching.

Can you kindly list, in both cases, the exact type of index, columns, dtype?

If this is not helping, I suppose we need to check whether the objects are identical, you can use `deep_equals`, or, more manually `dir`, `type`.

Abhay-Lejith commented:
@fkiraly Looking into obj deeper, I found that the the `blocks` were different in the two cases. (This is part of some internal pandas implementation). This was being caused by the way in which the predictions were being loaded into the resultant dataframe. 
I've fixed the issue now.

fkiraly commented:
How strange.

Would you mind explaining in more detail, how you detected and fixed this?
This seems indeed like a very hard to detect problem - typical `pandas` - so I think it is important we learn from this. FYI @sktime/core-developers.

Abhay-Lejith commented:
> Would you mind explaining in more detail, how you detected and fixed this?

I noticed that the traceback error was coming back from one of the pandas internal files. Looking at the code where it broke, it looked like it had something to do with the blocks not being of the correct shape, hence the broadcasting error which showed in CI ( I do not know what exactly this `block` is but it had something to do with how pandas stores data in the dataframe). 

So I used a breakpoint on the line where the code broke and examined `obj`. Vscode debugger shows full details, including these internal pandas attributes. The blocks were of different types in the two cases(the succcessful arima test and the failing skforecast test). The arrays stored in each block were also different. So I figured it had something to do with how the bootstrapped predictions were being loaded into the multiindexed result dataframe. 
Using breakpoints and debugging that part of the code, I discovered the error while loading the data and made the necessary modifications. (This was also the reason for the different quantile outputs being same, so thats fixed as well @yarnabrina ).

What's confusing is that though the blocks showed different data being stored within, printing `obj` gave the same output.

yarnabrina commented:
Based on commit 5688f93, changing the 1st one to 2nd one does the fix?

```python
quantile_predictions[(var_name, quantile)] = bootstrap_quantiles.to_numpy()[horizon_positions]
```

```python
quantile_predictions[(var_name, quantile)] = bootstrap_quantiles[quantile].iloc[horizon_positions.to_list()]
```

These two appear very different, the second one seems to have one aditional "column" selection corresponding to `quantile`. I wonder how are these being printed same.

---

I see that the 1st one is actually my old code from #5447 and there also I noted quantiles being same. I think this was an actual bug comprised of two issues: same quantile predictions were being stored for all specifications and column selection on dataframe to get series was not done.

fkiraly commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,451 @@
+from sktime.datatypes._utilities import get_cutoff
+from sktime.forecasting.base import BaseForecaster, ForecastingHorizon
+
+__author__ = ["yarnabrina"]
```
please add yourself as an author, @Abhay-Lejith

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,469 @@
+        ) = self._get_horizon_details(fh)
+
+        X_new = X
+        if X is not None:
```
The following check seems repetitive, can you address that please?

Abhay-Lejith replied:
This check is necessary here since X is optional and could be None.

yarnabrina replied:
I did not mean just L325. These three groups, L246-L264, L324-L336 and L401-L413 are very similar, and last two are probably identical.

Abhay-Lejith replied:
I can make a common function for them and call it in fit, predict and predict_quantile. Will that be fine?

yarnabrina replied:
Yes, let's have that as private method.

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,469 @@
+        )
+
+        X_new = X
+        if X is not None:
```
Anlther repetition

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -50,6 +49,13 @@ class SkforecastAutoreg(BaseForecaster):
+        residuals into k bins according to the predicted values associated with each
+        residual. The `encode' argument is always set to 'ordinal' and `dtype' to
+        np.float64.
+    forecaster_id : str, int, default `None`
```
I just added it because it was part of the skforecast args for ForecasterAutoreg. Should I remove it then?

yarnabrina replied:
Yes, please.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -410,10 +420,11 @@ def _predict_quantiles(
-
+            quantile_predictions[(var_name, quantile)] = (
+                bootstrap_quantiles[quantile]
+                .iloc[horizon_positions.to_list()]
```
Yes, it is necessary. horizon_positions is of index type. to_list makes it a regular list of integers which can be used in iloc.

yarnabrina replied:
Probably didn't understand you. The following works, can you please tell how is `horizon_positions` different?

```python
>>> import pandas
>>> a = pandas.Index([2, 4, 1, 5, 3])
>>> b = pandas.Series(range(10))
>>> b.iloc[a]
2    2
4    4
1    1
5    5
3    3
dtype: int64
>>> b.iloc[a.tolist()]
2    2
4    4
1    1
5    5
3    3
dtype: int64
>>> type(a)
<class 'pandas.core.indexes.base.Index'>
```

Abhay-Lejith replied:
Never mind, you are right. Though I remember using to_list() because of some indexing error, it works now without it as well. I will remove it.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,469 @@
+    @staticmethod
+    def _coerce_int_to_range_index(df):
+        new_df = df.copy(deep=True)
+        if new_df is not None:
```
Yes, I forgot to remove this check. Will change it.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,469 @@
+    def _coerce_int_to_range_index(df):
+        new_df = df.copy(deep=True)
+        if new_df is not None:
+            new_index = pd.RangeIndex(new_df.index[0], new_df.index[-1] + 1)
```
It will fail in skforecast. Integer index is not supported in skforecast.
Here it will fail saying could not coerce to RangeIndex.

yarnabrina replied:
Sorry, trying to understand a bit better.

Looking at this (https://skforecast.org/0.12.1/user_guides/input-data.html), it seems non-DatetimeIndex will be handled by skforecast directly. Are you saying that is not happening?

For example, this works for me albeit with warnings. With this PR's code, will the sktime wrapper work?

```python
import numpy
import pandas
from sklearn.linear_model import LinearRegression
from skforecast.ForecasterAutoreg import ForecasterAutoreg

random_generator = numpy.random.default_rng(seed=0)

sample_data = {
    "target_column": random_generator.standard_normal(size=10)
}
sample_index = pandas.Index(random_generator.permutation(10))
sample_dataset = pandas.DataFrame(data=sample_data, index=sample_index)

regressor = LinearRegression()
forecaster = ForecasterAutoreg(regressor, 1)

forecaster.fit(sample_dataset["target_column"])  # works
print(forecaster.fitted)  # True

forecaster.predict(5)  # works
```

Abhay-Lejith replied:
That's weird. In the earlier state of this PR, i.e #5447 , it would fail at the below point.
```
index = Index([49, 50, 51], dtype='int64'), steps = 1

    def expand_index(
        index: Union[pd.Index, None],
        steps: int
    ) -> pd.Index:
        """
        Create a new index of length `steps` starting at the end of the index.
    
        Parameters
        ----------
        index : pandas Index, None
            Original index.
        steps : int
            Number of steps to expand.
    
        Returns
        -------
        new_index : pandas Index
            New index.
    
        """
    
        if isinstance(index, pd.Index):
    
            if isinstance(index, pd.DatetimeIndex):
                new_index = pd.date_range(
                                start   = index[-1] + index.freq,
                                periods = steps,
                                freq    = index.freq
                            )
            elif isinstance(index, pd.RangeIndex):
                new_index = pd.RangeIndex(
                                start = index[-1] + 1,
                                stop  = index[-1] + 1 + steps
                            )
        else:
            new_index = pd.RangeIndex(
                            start = 0,
                            stop  = steps
                        )
    
>       return new_index
E       UnboundLocalError: local variable 'new_index' referenced before assignment
```
So integer and period indexes were not working. So I made the function to convert the index types.

Abhay-Lejith replied:
>Looking at this (https://skforecast.org/0.12.1/user_guides/input-data.html), it seems non-DatetimeIndex will be handled by skforecast directly. Are you saying that is not happening?

Not when we pass X. skforecast doesn't seem to convert the index type of X.  It fails with error saying index must be range or datetime.
Try this snippet passing X:
```python
import numpy
import pandas
from sklearn.linear_model import LinearRegression
from skforecast.ForecasterAutoreg import ForecasterAutoreg
from sktime.split import temporal_train_test_split

random_generator = numpy.random.default_rng(seed=0)

list=[1,2,3,4,5,6,7,8]
sample_data = list
sample_index = pandas.Index(list)
y = pandas.Series(data=sample_data, index=sample_index, name="y")

X = pandas.DataFrame(data=sample_data ,index=sample_index,columns=["X"])


y_train,y_test,X_train,X_test=temporal_train_test_split(y,X,0.5)
regressor = LinearRegression()
forecaster = ForecasterAutoreg(regressor, 1)

forecaster.fit(y_train,X_train)  
print(forecaster.fitted) 
forecaster.predict(4,exog=X_test)
```

> With this PR's code, will the sktime wrapper work?

I didn't understand, this PR's code is working right? Otherwise how are all tests passing?

yarnabrina replied:
I'll try with X tonight after office, will revert.

As of now, looking at the code I think just y with non-sequential index (similar to my snippet above) will fail with your adapter class, even though it works directly with skforecast. Can you please confirm if that is my misunderstanding?

> Otherwise how are all tests passing?

I am not sure if sktime's default testing suite contains test case with non-sequential integers being the index or similar. May be @fkiraly can confirm on that, but we can also add a new test case.

Abhay-Lejith replied:
> As of now, looking at the code I think just y with non-sequential index (similar to my snippet above) will fail with your adapter class.

Yes, if Integer index is passed and it is non-sequential. It will fail here because it won't be converted to RangeIndex correctly. 
This is a consequence of the index type conversion that I am doing in order to support all index types in exog. I will see if I can figure out a way to convert non-sequential index correctly.
Or, I can make the index conversion only when `X` is passed. So in your example above, since no `X` is passed, I don't convert index type and it will run without error. Only downside with this is that non-sequential index in `X` and `y` will not work. But this does not work in skforecast either.

> I am not sure if sktime's default testing suite contains test case with non-sequential integers being the index or similar.

If I'm not wrong, I think all fixtures have sequential indices.

Abhay-Lejith replied:
@yarnabrina I checked with some other models, non-sequential indices do not work in sktime. I was getting errors saying that index must be monotonically increasing.

yarnabrina replied:
> Not when we pass X. It fails with error saying index must be range or datetime.

Yes, I am facing that only for prediction, training seems to happen fine. Interestingly, even with range index I am facing the same error.

```python
import numpy
import pandas
from skforecast.ForecasterAutoreg import ForecasterAutoreg
from sklearn.linear_model import LinearRegression

random_generator = numpy.random.default_rng(seed=0)

sample_data = {
    "target_column": random_generator.standard_normal(size=10),
    "numeric_column": random_generator.standard_normal(size=10),
}
sample_index = pandas.Index(random_generator.permutation(10))
sample_dataset = pandas.DataFrame(data=sample_data, index=sample_index)

regressor = LinearRegression()
forecaster = ForecasterAutoreg(regressor, 1)

forecaster.fit(sample_dataset.iloc[:7, 0], exog=sample_dataset.iloc[:7, 1])  # works
print(forecaster.fitted)  # True

forecaster.predict(3, exog=sample_dataset.iloc[7:, 1])  # fails

hacky_future_exog_dataset = sample_dataset.iloc[7:, 1]
hacky_future_exog_dataset.index = pandas.RangeIndex(start=7, stop=10, step=1)

forecaster.predict(3, exog=hacky_future_exog_dataset)  # fails (why?)
```

Regarding how should we fail, I am thinking if we can use some tags for this purpose. A tag can be used to show if a particular estimator has the capability to forecast in presence of gaps in the data, and for all such cases we can fail with a common error. @fkiraly / @benHeid any thoughts?

As of now, the only suggestion I have is to allow step to be something other than 1, everything else seems fine. The way you are creating `new_index` implies `step=1` always, can we allow that as the gap between positions 0 and 1?

Abhay-Lejith replied:
Yes, we are assuming step=1 here. I guess we can also make step= index[1]-index[0] but this would assume that the indexes are equally spaced apart. 
If the integer index isn't spaced equally apart, I don't see any solution as it cannot be converted to any supported index type.

yarnabrina replied:
Right, and we'll keep it as it is. So, for now, I suggest you make two changes: allow constant step (but not necessarily 1) and in the error message inform whether it's failing for `y` or `X`.

We did a similar type of thing in `neuralforecast` adapter earlier, if I remember correctly. Not on laptop at the moment to check, but if possible try to use similar error message if possible.

Abhay-Lejith commented:
@yarnabrina I have addressed all the issues discussed. This is ready for review now.

Abhay-Lejith commented:
@yarnabrina I've also done one of the TODO's you had listed in #5447 which was to modify predict quantiles to use predict_quantiles from skforecast instead of the bootstrapping+np.quantile . Tests for the adapter have also been added.

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,445 @@
+    def _coerce_period_to_datetime_index(df):
+        new_df = df.copy(deep=True)
+        period_freq = new_df.index.freq
+        new_df.index = new_df.index.to_timestamp(how="e").normalize()
```
It'll be great if you make a comment here explaining that here we are converting period indices to the end date.

Curious, can period be sub-daily? How does this behave then?

Abhay-Lejith replied:
I just checked, it can be hourly and so on. I've removed the normalize() call so that it works now.

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,445 @@
+                    "pd.RangeIndex, pd.DatetimeIndex, pd.PeriodIndex, pd.Index"
+                    f"(int dtype). Found index of type: {type(df.index)}"
+                )
+        return new_df
```
This is coresponding to range or datetime indices, correct? To reduce nesting, which increases mccabe complexity, I suggest you make a positive check for those first and return `df` as it is. And then proceed to nested checks, without nesting this time.

For successful coercion case, use try-except-else or similar.

yarnabrina reviewed (changes requested):
Great work, almost done. Added few comments.

Abhay-Lejith commented on sktime/forecasting/compose/tests/test_skforecast_reduce.py:
skforecast model doesn't support multivariate itself, thats why I didn't add it.

yarnabrina replied:
It does have https://joaquinamatrodrigo.github.io/skforecast/0.12.1/api/ForecasterMultiVariate.html, may be that can be the future scope to support that?

Abhay-Lejith replied:
Hmm, it is a different model though, not `ForecasterAutoreg`. But yes, we can extend to support multivariate in a separate issue/pr.

Abhay-Lejith commented on sktime/forecasting/compose/tests/test_skforecast_reduce.py:
the dataframe shapes and column names arent same in most cases, thats why I have chosen to just check the values in the dataframe.

Abhay-Lejith replied:
Never mind, there are arguments to not check column names and similar in `assert_frame_equal` , I will use it as suggested.

Abhay-Lejith commented on sktime/forecasting/compose/tests/test_skforecast_reduce.py:
```diff
@@ -0,0 +1,123 @@
+
+    y, X = make_forecasting_problem(n_timepoints=10, make_X=True, index_type="range")
+    fh = [1, 2, 3]
+    X.columns = X.columns.astype("str")
```
> If it's to pass to skforecast, I suggest to make the change for that only
Okay, will do it this way.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,445 @@
+        new_df = df.copy(deep=True)
+        start = new_df.index[0]
+        stop = new_df.index[-1] + 1
+        if len(new_df.index) == 1:
```
To find step I am doing index[1]-index[0], If the length of index were 1, accessing index[1] would throw an error so I separately specified step=1. It technically would not matter as when index length is 1, there is only a single index value so step is not used.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,445 @@
+            step = new_df.index[1] - start
+
+        new_index = pd.RangeIndex(start, stop, step)
+        np.testing.assert_array_equal(new_df.index, new_index)
```
I am catching the error and giving an informative error message in `_make_index_compatible`.

yarnabrina replied:
I'm not really a fan, say from readability point of view. If I read this code, it looks like I am checking if B matches A, and if it does, replace A with B, which should have no effect.

Does that make sense?

Abhay-Lejith replied:
I understand. We are indeed checking if B matches A, but only in value. 
However both are of different index types which is why we are replacing one with the other. i.e. to convert to the compatible index type.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,445 @@
+                f"{self.__class__.__name__} does not support in-sample predictions."
+            )
+
+        if not (out_of_sample_horizon := fh.to_out_of_sample(self.cutoff)):
+            raise ValueError(
+                f"{self.__class__.__name__} received empty out-of-sample horizon."
+            )
```
This part was written by you actually. I am not very familiar with fh, but I think it does look like a redundant check. The previous if statement ensures that fh is 'all out of sample' so `out_of_sample_horizon` cannot never be None here. Is this correct?

yarnabrina replied:
I also felt it was redundant, don't remember why I wrote that. Let me see my local branch internal commit history.

yarnabrina replied:
Is it enough reason to keep this check?

```python
>>> 
>>> ForecastingHorizon([]).is_all_in_sample()
True
>>> ForecastingHorizon([]).is_all_out_of_sample()
True
>>> 
```

I am not too convinced, but will probably prefer to keep as it is since it's not doing any harm.

Abhay-Lejith replied:
I did actually test this case, I found that sktime doesn't allow empty fh to be passed to the estimator and error is raised beforehand itself. 
So I would say it's safe to remove.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,445 @@
+
+        for quantile in alpha:
+            quantile_predictions[(var_name, quantile)] = (
+                quantile_pred[f"q_{quantile}"].iloc[horizon_positions].to_list()
```
Any reason for why you prefer one over the other? I had a look online and it seems they both are the exact same function, just aliases.

yarnabrina replied:
Personal preference, really. `to_list` is an alias of `tolist`, and `tolist` uses `numpy.ndarray.tolist`.

Abhay-Lejith commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,445 @@
+        if len(new_df.index) == 1:
+            step = 1
+        else:
+            step = new_df.index[1] - start
```
This would also work. I have explained what I have done above in the other comment.

yarnabrina commented on sktime/forecasting/compose/_skforecast_reduce.py:
```diff
@@ -0,0 +1,452 @@
+        if isinstance(df.index, (pd.RangeIndex, pd.DatetimeIndex)):
+            return df
+
+        new_df = df
```
Is this assignment used?

Abhay-Lejith replied:
No, it is not necessary now. I've removed it.

yarnabrina reviewed (approved):
Made few comments, but I am mostly happy with this merge request. Approved.

Abhay-Lejith commented:
Failing test looks like an unrelated macos packages error.

fkiraly commented:
I don't see a failing test?

Abhay-Lejith commented:
> I don't see a failing test?

Neither do I now. Weird. There certainly was a macos packages hash not matching error earlier.

fkiraly commented on docs/source/api_reference/forecasting.rst:
```diff
@@ -469,6 +469,14 @@ Generic framework adapters that expose other frameworks in the ``sktime`` interf
+    :toctree: auto_generated/
+    :template: class.rst
+
+    SkforecastAutoreg
```
I think the right category is "reduction forecasters".

I would also add a short summary at the top on the three types of reduction forecasters available, and sort them into subcategories now that there are three bunches.

* 1st generation `sktime` native, via `make_reduction`
* 2nd generation `sktime` native, `DirectReductionForecaster` and `RecursiveReductionForecaser`
* from `skforecast`, `SkforecastAutoreg`

Abhay-Lejith replied:
Sure. Could you please tell what the names of the three subcategories need to be? I don't really get the meaning of 1st/2nd generation.

fkiraly replied:
I'll change it

fkiraly reviewed (changes requested):
May I request to sort the new forecaster onto the "right shelf"?

See above. I think there are now enough to split in three subcategories.

yarnabrina commented:
> > I don't see a failing test?
> 
> Neither do I now. Weird. There certainly was a macos packages hash not matching error earlier.

I restarted the failed CI earlier, so it's gone now. It was the common SHA256 mismatch error in pip install step.

If you want, this is the failed job: https://github.com/sktime/sktime/actions/runs/9723394102/job/26838541739#step:3:360.

fkiraly reviewed (dismissed):
Updated the API ref

JavierEscobarOrtiz commented:
Hello @fkiraly @Abhay-Lejith @yarnabrina 

I’m Javier, one of the co-authors of skforecast. First of all, thank you so much for including our library as a wrapper in sktime. It’s amazing to see how our work helps other teams!

I wanted to let you know that in skforecast 0.14.0, we’ve implemented a major refactoring aimed at improving performance and usability. For instance, the class ForecasterAutoreg has been renamed to ForecasterRecursive:

```python
# Before Skforecast 0.14
from skforecast.ForecasterAutoreg import ForecasterAutoreg

# From skforecast 0.14.0 onward 
from skforecast.recursive import ForecasterRecursive
```

Same Forecaster but new name and new features.

If you need help with the integration, just ping us! 😄 We have also created a [migration guide](https://skforecast.org/0.14.0/user_guides/migration-guide) section for anyone who wants to update to the de version.

Best,

Javi, @JoaquinAmatRodrigo
